internal/locate/region_request.go

```diff
@@ -1388,6 +1402,48 @@ func (s *RegionRequestSender) SendReqCtx(
 		}()
 	}
 
+	type backoffArgs struct {
+		cfg *retry.Config
+		err error
+	}
+	pendingBackoffs := make(map[string]*backoffArgs)
+	delayBoTiKVServerBusy := func(addr string, reason string) {
+		if _, ok := pendingBackoffs[addr]; !ok {
+			pendingBackoffs[addr] = &backoffArgs{
+				cfg: retry.BoTiKVServerBusy,
+				err: errors.Errorf("server is busy, reason: %s", reason),
+			}
+		}
+	}
+	backoffOnRetry := func(addr string) error {
+		if args, ok := pendingBackoffs[addr]; ok {
+			delete(pendingBackoffs, addr)
+			logutil.Logger(bo.GetCtx()).Warn(
+				"apply pending backoff on retry",
+				zap.String("target", addr),
+				zap.String("bo", args.cfg.String()),
+				zap.String("err", args.err.Error()))
+			return bo.Backoff(args.cfg, args.err)
+		}
+		return nil
+	}
+	backoffOnFakeErr := func() error {
+		var args *backoffArgs
+		// if there are multiple pending backoffs, choose the one with the largest base duration.
+		for _, it := range pendingBackoffs {
+			if args == nil || args.cfg.Base() < it.cfg.Base() {
+				args = it
+			}
+		}
+		if args != nil {
+			logutil.Logger(bo.GetCtx()).Warn(
+				"apply pending backoff on pseudo region error",
+				zap.String("bo", args.cfg.String()),
+				zap.String("err", args.err.Error()))
+			return bo.Backoff(args.cfg, args.err)
+		}
+		return nil
+	}
```

Comment:
Could you explain the purpose of these code? I didn't understand it 🤔

Reply:
It's used to maintain delayed backoffs.
-  "`pendingBackoffs[addr]` is not nil" means there is a delayed backoff which should be applied on retrying this store.
- `delayBoTiKVServerBusy` is the callback for recording delayed backoffs.
- `backoffOnRetry` is used to apply pending backoffs on retry.
- when there is no candidate, the kv client just return a fake error and let the caller do retry. to avoid potential frequent rpcs, we call `backoffOnFakeErr` before returning a fake error, which chooses a pending backoff with the largest base duration and applies it.

Reply:
Can this be done by adding "busy" flag to the stores, instead of maintaining maps here?

Reply:
If you mean the [store](https://github.com/tikv/client-go/blob/dbea404a846b5404206f1455da4e94639da9895f/internal/locate/region_cache.go#L2535) struct in region cache, probably no I think. They have different scope/lifetime, that is, each `SendReqCtx` call should has its own `pendingBackoffs`, and has no effect to each others. Maybe we can add a field to the [replica](https://github.com/tikv/client-go/blob/master/internal/locate/region_request.go#L247) struct, however, this PR is just a quick-and-dirty fix to the customer issue, I just try to make the least change.